klaas/scripts/apply_separation_model.py

```diff
@@ -1,74 +1,111 @@
+            click.confirm(
+                'Dataset "signal_prediction" exists in file, overwrite?',
+                abort=True,
+            )
+        del f[key]['signal_prediction']
+        for region in range(1, 6):
+            dataset = 'background_prediction_{}'.format(region)
+            if dataset in f[key].keys():
+                del f[key][dataset]
+
+    log.info('Loading model')
     model = joblib.load(model_path)
-    #sklearn needs float32 values. after downcasting some -infs appear somehow. here i drop them.
-    print('Loading data')
-    df_data = read_data(data_path, query=query)
-    df_data[training_variables] = df_data[training_variables].astype('float32')
-    df_data = df_data.replace([np.inf, -np.inf], np.nan).dropna(how='any')
-
-
-    # embed()
-    print('After dropping nans there are {} events left.'.format(len(df_data)))
-    print('Predicting on data...')
-    prediction = model.predict_proba(df_data[training_variables])
-    df_data['signal_prediction'] = prediction[:,1]
-    df_data['signal_theta'] = df_data['Theta']
-    df_data['signal_distance'] = df_data['Distance']
-    if 'Theta' in training_variables:
-        thetas = df_data['Theta'].copy()
-        distances = df_data['Distance'].copy()
-        alphas = df_data['Alpha'].copy()
-
-        print('Predicting off data...')
-        for region in [1,2,3,4,5]:
-            theta_key = 'Theta_Off_{}'.format(region)
-            distance_key = 'Distance_Off_{}'.format(region)
-            alpha_key = 'Alpha_Off_{}'.format(region)
-            df_data['Theta'] = df_data[theta_key]
-            df_data['Distance'] = df_data[distance_key]
-            df_data['Alpha'] = df_data[alpha_key]
-            prediction = model.predict_proba(df_data[training_variables])
-            df_data['background_prediction_{}'.format(region)] =  prediction[:,1]
-
-
-        df_data['Distance'] = distances
-        df_data['Theta'] = thetas
-        df_data['Alphas'] = alphas
-
-    print('Writing data')
-    write_data(df_data, output_path)
-
+    log.info('Done')
+
+    used_source_features = find_used_source_features(training_variables)
+    if len(used_source_features) > 0:
+        log.info(
+            'Source dependent features used in model, '
+            'redoing classification for off regions'
+        )
+
+    needed_features = [
+        var + '_Off_{}'.format(region)
+        for region in range(1, 6)
+        for var in used_source_features
+    ]
+
+    df_generator = read_h5py_chunked(
+        data_path,
+        key=key,
+        columns=training_variables + needed_features,
+        chunksize=chunksize,
+    )
+
+    log.info('Predicting on data...')
+    for df_data, start, end in tqdm(df_generator):
+
+        signal_prediction = predict(df_data, model, training_variables)
+
+        with h5py.File(data_path) as f:
+            if 'signal_prediction' in f[key].keys():
+                n_existing = f[key]['signal_prediction'].shape[0]
```

Comment:
The code is from the docs, basically :(

http://docs.h5py.org/en/latest/faq.html#appending-data-to-a-dataset